spec/unit/iiif/v3/presentation/service_spec.rb

```diff
@@ -1,28 +1,157 @@
+      expect(service_obj['service'][0]['profile']).to eq inner_service_profile
     end
-    context "when there are singleton classes which are returned" do
-      it "gets the right class" do
-        allow(IIIF::V3::Presentation::Service).to receive(:descendants).and_return([DummyClass, IIIF::V3::Presentation::Collection])
-        klass = described_class.get_descendant_class_by_jld_type('Collection')
-        expect(klass).to eq IIIF::V3::Presentation::Collection
+    it 'allows both "id" and "@id" as keys' do
+      id_uri = "https://stacks.stanford.edu/image/iiif/wy534zh7137%2FSULAIR_rosette"
+      service_obj = described_class.new({
+        'id' => id_uri,
+        '@id' => id_uri
+      })
+      expect(service_obj.keys.size).to eq 2
+      expect(service_obj['id']).to eq id_uri
+      expect(service_obj['@id']).to eq id_uri
+    end
+    it 'allows non-URI profile value' do
+      expect{
+        described_class.new({
+          "profile" => [
+            "http://iiif.io/api/image/2/level2.json",
+            {
+              "formats" => [ "gif", "pdf" ],
+              "qualities" => [ "color", "gray" ],
+              "supports" => [ "canonicalLinkHeader", "rotationArbitrary", "http://example.com/feature" ]
+            }
+          ]
+        })
+      }.not_to raise_error
+    end
+  end
+
+  describe '#validate' do
+    describe '@context = IIIF_IMAGE_API_V2_CONTEXT' do
+      it 'must have a "@id"' do
+        service_obj = described_class.new({
+          '@context' => described_class::IIIF_IMAGE_V2_CONTEXT,
+          'id' => id_uri,
+          'profile' => described_class::IIIF_IMAGE_V2_LEVEL1_PROFILE
+          })
+        exp_err_msg = '@id is required for IIIF::V3::Presentation::Service with @context http://iiif.io/api/image/2/context.json'
+        expect{service_obj.validate}.to raise_error(IIIF::V3::Presentation::MissingRequiredKeyError, exp_err_msg)
+      end
+      it 'must have a profile' do
+        service_obj = described_class.new({
+          '@context' => described_class::IIIF_IMAGE_V2_CONTEXT,
+          '@id' => id_uri
+          })
+        exp_err_msg = 'profile is required for IIIF::V3::Presentation::Service with @context http://iiif.io/api/image/2/context.json'
+        expect{service_obj.validate}.to raise_error(IIIF::V3::Presentation::MissingRequiredKeyError, exp_err_msg)
+      end
+      it 'must have matching values for "@id" and "id" if both are specified' do
+        service_obj = described_class.new({
+          '@context' => described_class::IIIF_IMAGE_V2_CONTEXT,
+          '@id' => id_uri,
+          'id' => "#{id_uri}/foo"
+          })
+        exp_err_msg = 'id and @id values must match for IIIF::V3::Presentation::Service with @context http://iiif.io/api/image/2/context.json'
+        expect{service_obj.validate}.to raise_error(IIIF::V3::Presentation::IllegalValueError, exp_err_msg)
       end
     end
   end
 
+  describe '"integration" tests' do
```

Comment:
should this be in the `spec/integration` folder?  it doesn't use any fixtures, and it doesn't call out to anything ...